packages/compass-aggregations/src/modules/pipeline-builder/pipeline-parser/stage-parser.ts

```diff
@@ -18,32 +18,72 @@ export function setNodeDisabled(node: t.Node, value: boolean) {
   (node as any)[kDisabled] = value;
 }
 
-export function isStageLike(node?: t.Node | null, loose = false): node is StageLike {
-  return (
-    !!node &&
-    node.type === 'ObjectExpression' &&
-    node.properties.length === 1 &&
-    node.properties[0].type === 'ObjectProperty' &&
-    node.properties[0].key &&
-    ((node.properties[0].key.type === 'Identifier' &&
-      node.properties[0].key.name.startsWith('$')) ||
-      (node.properties[0].key.type === 'StringLiteral' &&
-        node.properties[0].key.value.startsWith('$'))) &&
-    (loose || node.properties[0].value != null)
-  );
+export function isValidStageNode(node?: t.ObjectExpression): boolean {
+  if (node) {
+    try {
+      // TODO: either export ejson-shell-parser logic that validates the AST or
+      // move the logic in this package to avoid generating source from ast
+      // before validating it again
+      parseEJSON(generate(node));
+      return true;
+    } catch (err) {
+      return false;
+    }
+  }
+  return false;
 }
 
-export function assertStageNode(node: t.Node): asserts node is StageLike {
-  if (isStageLike(node)) {
-    return;
+export function isStageLike(
+  node?: t.Node | null,
+  loose = false
+): node is StageLike {
+  try {
+    assertStageNode(node, loose);
+    return true;
+  } catch {
+    return false;
   }
-  const message = node.type === 'ObjectExpression'
-    ? (node.properties[0] as t.ObjectProperty | undefined)?.key == null
-      ? 'A pipeline stage specification object must contain exactly one field.'
-      : 'Stage value can not be empty'
-    : 'Each element of the pipeline array must be an object';
+}
 
-  throw new PipelineParserError(message, node.loc?.start);
+function getKeyName(node: t.ObjectProperty['key']): string | null {
+  return node.type === 'Identifier'
+    ? node.name
+    : node.type === 'StringLiteral'
+    ? node.value
+    : null;
+}
+
+export function assertStageNode(
+  node?: t.Node | null,
+  loose = false
+): asserts node is StageLike {
+  let error: string | null = null;
+  let causedBy = node;
+
+  if (!node || node.type !== 'ObjectExpression') {
+    error = 'Each element of the pipeline array must be an object';
+  } else if (
+    node.properties.length !== 1 ||
+    node.properties[0].type !== 'ObjectProperty' ||
+    node.properties[0].key == null
+  ) {
+    error =
+      'A pipeline stage specification object must contain exactly one field.';
+    causedBy = node.properties[0] ?? causedBy;
+  } else if (!getKeyName(node.properties[0].key)?.startsWith('$')) {
+    const key = getKeyName(node.properties[0].key) ?? '';
+    error = `Unrecognized pipeline stage name${key ? `: '${key}'` : ''}`;
+    causedBy = node.properties[0].key;
```

Comment:
The messages are a bit more detailed now, and we also track what part of the stage caused them so that the error annotations can be a bit more precise:

|Before|After|
|---|---|
|![image](https://user-images.githubusercontent.com/5036933/202690758-e76540d8-f80d-4d81-9165-77a3b75ad03a.png)|![image](https://user-images.githubusercontent.com/5036933/202690635-01fb369a-3ab6-47a5-b96e-d7a2208ae451.png)|